internal/receiver/discoveryreceiver/correlation.go

```diff
@@ -0,0 +1,218 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package discoveryreceiver
+
+import (
+	"sync"
+	"time"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/extension/observer"
+	"go.opentelemetry.io/collector/config"
+	"go.uber.org/zap"
+
+	"github.com/signalfx/splunk-otel-collector/internal/receiver/discoveryreceiver/statussources"
+)
+
+// correlation is a grouping of an endpoint, the observing
+// observer, and a receiver, if any. It's used to unify
+// emitted log record content from evaluated status sources
+// such that no individual source lacks information otherwise
+// only available to others. At this time this means that
+// metrics and component logs will have access to a receiver's
+// observing observer via the Notify event where not available
+// through another means.
+type correlation struct {
+	lastState   endpointState
+	lastUpdated time.Time
+	endpoint    observer.Endpoint
+	receiverID  config.ComponentID
+	observerID  config.ComponentID
+}
+
+// correlationStore provides a centralized interface for up-to-date correlations
+// and receiver attributes as a message passing mechanism by observed components.
+// It manages a reaping loop to prevent stale endpoint buildup over time.
+type correlationStore interface {
+	UpdateEndpoint(endpoint observer.Endpoint, state endpointState, observerID config.ComponentID)
+	GetOrCreate(receiverID config.ComponentID, endpointID observer.EndpointID) correlation
+	Attrs(receiverID config.ComponentID) map[string]string
+	UpdateAttrs(receiverID config.ComponentID, attrs map[string]string)
+	// Start the reaping loop to prevent unnecessary endpoint buildup
+	Start()
+	// Stop the reaping loop
+	Stop()
+}
+
+// store is a collection of mappings used as an instantaneous record of
+// 1. endpoints to their associated receivers->correlations
+// 2. receivers to their endpoint-agnostic Attrs used as a message
+// passing mechanism (currently just for embedded config values).
+// This way the pre-created receiver instances can have attrs
+// before they are created via endpoint.
+type store struct {
+	logger *zap.Logger
+	// correlations is a ~synchronized map[endpointID]map[receiverID]*corr
+	correlations  *sync.Map
+	endpointLocks *keyLock
+	receiverAttrs *sync.Map
+	receiverLocks *keyLock
+	// sentinel for terminating reaper loop
+	sentinel     chan struct{}
+	reapInterval time.Duration
+	ttl          time.Duration
+}
+
+func newCorrelationStore(logger *zap.Logger, ttl time.Duration) correlationStore {
+	return &store{
+		logger:        logger,
+		correlations:  &sync.Map{},
+		endpointLocks: newKeyLock(),
+		receiverAttrs: &sync.Map{},
+		receiverLocks: newKeyLock(),
+		reapInterval:  30 * time.Second,
```

Comment:
Do we care if the ttl is lower than reapInterval? Should reapInterval be half of ttl?

Reply:
For high churn environments adding the reap interval as another configuration option may make sense with much shorter ttls but I opted not to do this out of the gate to reduce configuration complexity. As I see it general use case of the reaper is for `log_endpoints` or a few debug/context specific receivers on long lived collector instances to ensure that cycling endpoints don't cause leaks over the long run (10min default ttl w/ hardcoded 30s reap interval).

I think the other edge cases outside of heavy observer activity involve programming errors in receivers that fail to halt metric or app log generation after they've been halted.

Reply:
10 mins ttl default sounds good to me. OK!